system/lib/pthread/proxying.h

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2021 The Emscripten Authors.  All rights reserved.
+ * Emscripten is available under two separate licenses, the MIT license and the
+ * University of Illinois/NCSA Open Source License.  Both these licenses can be
+ * found in the LICENSE file.
+ */
+
+#pragma once
+
+#include <emscripten/emscripten.h>
+#include <pthread.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+// Opaque handle to a set of thread-local work queues to which work can be
+// asynchronously or synchronously proxied from other threads. When work is
+// proxied to a queue on a particular thread, that thread is notified to start
+// processing work from that queue if it is not already doing so.
+typedef struct em_proxying_queue em_proxying_queue;
+
+// Create and destroy proxying queues.
+em_proxying_queue* em_proxying_queue_create();
+void em_proxying_queue_destroy(em_proxying_queue* q);
+
+// Get the queue used for proxying low-level runtime work. Work on this queue
+// may be processed at any time inside system functions, so it must be
+// nonblocking and safe to run at any time, similar to a native signal handler.
+em_proxying_queue* emscripten_proxy_get_system_queue();
+
+// Execute all the tasks enqueued for the current thread on the given queue.
+// Exported for use in worker.js.
+EMSCRIPTEN_KEEPALIVE
+void emscripten_proxy_execute_queue(em_proxying_queue* q);
```

Comment:
Does this get emitted to all builds, even if user does not use this proxying queue? This should get gated behind a `#if LIBRARY_PROXY_QUEUE` if so.

Reply:
The way I was suggesting this should work is that if any function from `system/lib/pthread/proxying.c` is included in the link then this symbol will be exported for JS so the JS can call it (that is how EMSCRIPTEN_KEEPALIVE works).

If there is no dependency on `proxying.c` at all then nothing should be exported.   This way the JS code and use `hasExportedFunction('emscripten_proxy_execute_queue')` and know that it will be exported if and only if the program makes some use of proxying.

Reply:
@tlively can you put `EMSCRIPTEN_KEEPALIVE` instead on the defintion in the C file?  (or does that not work?)

Reply:
Yes, I can make that change.